File: android/src/main/kotlin/com/pubmatic/sdk/flutter_openwrap_sdk/interstitial/POBInterstitialClient.kt
package com.pubmatic.sdk.flutter_openwrap_sdk.interstitial

import android.content.Context
import com.pubmatic.sdk.common.POBError
import com.pubmatic.sdk.flutter_openwrap_sdk.POBAdClient
import com.pubmatic.sdk.flutter_openwrap_sdk.POBUtils
import com.pubmatic.sdk.openwrap.interstitial.POBInterstitial
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Wrapper around [POBInterstitial] to transfer dart method calls to OpenWrap SDK.
 *
 * @property channel [MethodChannel] for native-dart communication
 * @constructor
 * Initializes the [POBInterstitial] with necessary properties
 *
 * @param adId Unique Instance manager id
 * @param channel MethodChannel for triggering callback on dart side
 * @param context Application context
 * @param pubId Identifier of the publisher
 * @param profileId Profile ID of an ad tag
 * @param adUnitId Ad unit id used to identify unique placement on screen
 * @param isHeaderBidding Boolean indicates whether the GAM flow is included or not
 */
class POBInterstitialClient(
  adId: Int,
  channel: MethodChannel,
  context: Context,
  pubId: String,
  profileId: Int,
  adUnitId: String,
  isHeaderBidding: Boolean
) : POBAdClient(adId, channel) {

  private val interstitial: POBInterstitial
  private var eventHandler: POBFLTInterstitialEvent? = null

  // Initialize [POBInterstitialDefaultEventHandler] in case of header bidding flow
  // by using [isHeaderBidding] parameter else use the default constructor
  init {
    if (isHeaderBidding) {
      eventHandler = POBInterstitialEventHandlerClient(adId, channel)
      interstitial = POBInterstitial(context, pubId, profileId, adUnitId, eventHandler!!)
    } else {
      interstitial = POBInterstitial(context, pubId, profileId, adUnitId)
    }
    request = interstitial.adRequest
    impression = interstitial.impression
  }

  override fun methodCall(names: List<String>, call: MethodCall, result: Result) {
    when (names[1]) {
      "loadAd" -> {
        interstitial.loadAd()
        result.success(null)
      }

      "show" -> {
        interstitial.show()
        result.success(null)
      }

      "isReady" -> result.success(interstitial.isReady)


      "setListener" -> {
        interstitial.setListener(POBInterstitialListenerImpl(channel, adId))
        result.success(null)
      }

      "setVideoListener" -> {
        interstitial.setVideoListener(POBVideoListenerImpl(channel, adId))
        result.success(null)
      }

      "getBid" -> result.success(convertBidToMap(interstitial.bid))

      "destroy" -> {
        super.destroy()
        interstitial.destroy()
        result.success(null)
      }

      "EventHandler" -> eventHandler?.methodCall(names[2], call, result)

      else -> super.methodCall(names, call, result)
    }
  }

  /**
   * Listener for receiving callbacks of [POBInterstitial] events from OpenWrapSDK.
   */
  private class POBInterstitialListenerImpl(
    private val channel: MethodChannel,
    private val adId: Int
  ) : POBInterstitial.POBInterstitialListener() {

    override fun onAdReceived(ad: POBInterstitial) {
      channel.invokeMethod("onAdReceived", POBUtils.getArgumentMap(adId))
    }

    override fun onAdFailedToLoad(ad: POBInterstitial, error: POBError) {
      channel.invokeMethod("onAdFailedToLoad", POBUtils.getArgumentMap(adId, error))
    }

    override fun onAdFailedToShow(ad: POBInterstitial, error: POBError) {
      channel.invokeMethod("onAdFailedToShow", POBUtils.getArgumentMap(adId, error))
    }

    override fun onAppLeaving(ad: POBInterstitial) {
      channel.invokeMethod("onAppLeaving", POBUtils.getArgumentMap(adId))
    }

    override fun onAdOpened(ad: POBInterstitial) {
      channel.invokeMethod("onAdOpened", POBUtils.getArgumentMap(adId))
    }

    override fun onAdClosed(ad: POBInterstitial) {
      channel.invokeMethod("onAdClosed", POBUtils.getArgumentMap(adId))
    }

    override fun onAdClicked(ad: POBInterstitial) {
      channel.invokeMethod("onAdClicked", POBUtils.getArgumentMap(adId))
    }

    override fun onAdExpired(ad: POBInterstitial) {
      channel.invokeMethod("onAdExpired", POBUtils.getArgumentMap(adId))
    }

    override fun onAdImpression(ad: POBInterstitial) {
      channel.invokeMethod("onAdImpression", POBUtils.getArgumentMap(adId))
    }
  }

  /**
   * Listener for receiving callbacks of [POBInterstitial] video events from OpenWrapSDK.
   */
  private class POBVideoListenerImpl(
    private val channel: MethodChannel,
    private val adId: Int
  ) : POBInterstitial.POBVideoListener() {
    override fun onVideoPlaybackCompleted(ad: POBInterstitial) {
      channel.invokeMethod("onVideoPlaybackCompleted", POBUtils.getArgumentMap(adId))
    }
  }
}

File: ios/Classes/POBInterstitialClient.swift
import UIKit
import OpenWrapSDK
import Flutter

/// Wrapper class around `POBInterstitial` to transfer the API calls
class POBInterstitialClient: POBAdClient, POBInterstitialDelegate, POBInterstitialVideoDelegate {
    var instl: POBInterstitial?
    var methodChannel: FlutterMethodChannel?
    var eventHandler: POBFLTInterstitialEvent?

    /// Initiates a new interstitial client
    /// - Parameters:
    ///
    /// - adIndex: index to identify the client instance
    /// - pubId: Identifier of the publisher
    /// - profileId: OpenWrap profile ID
    /// - adUnitId: Ad unit id used to identify unique placement on screen
    /// - channel: `FlutterMethodChannel` for native-dart communication
    init(adIndex: Int, publisherId: String, profileId: NSNumber,
         adUnitId: String, isHeaderBidding: Bool, channel: FlutterMethodChannel) {
        super.init(adId: adIndex)
        methodChannel = channel
        if isHeaderBidding {
            eventHandler = POBInterstitialEventHandlerClient(methodChannel: methodChannel, adId: adId)
            instl = POBInterstitial(publisherId: publisherId,
                                    profileId: profileId,
                                    adUnitId: adUnitId,
                                    eventHandler: eventHandler!)
        } else {
            instl = POBInterstitial(publisherId: publisherId,
                                    profileId: profileId,
                                    adUnitId: adUnitId)
        }
        request = instl?.request
    }
    
    override func invokeMethod(names: [String], call: FlutterMethodCall, result: FlutterResult) {
        switch names[1] {
        case "loadAd":
            instl?.loadAd()
            result(nil)
        case "show":
            instl?.show(from: UIApplication.presentedViewController())
            result(nil)
        case "isReady":
            result(instl?.isReady)
        case "setListener":
            instl?.delegate = self
            result(nil)
        case "setVideoListener":
            instl?.videoDelegate = self
            result(nil)
        case "getBid":
            result(convertBidToMap(bid: instl?.bid()))
        case "destroy":
            // ad is removed from super class switch case
            instl = nil
            result(nil)
        case "EventHandler":
            eventHandler?.invokeMethod(methodName: names[2], call: call, result: result)
        default:
            super.invokeMethod(names: names, call: call, result: result)
        }
    }
    
    // MARK: Interstitial delegate methods
    
    // Notifies the delegate that an ad has been received successfully.
    func interstitialDidReceiveAd(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdReceived", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate of an error encountered while loading or rendering an ad.
    func interstitial(_ interstitial: POBInterstitial, didFailToReceiveAdWithError error: Error) {
        let errMap = POBUtils.convertErrorAndAdIdToMap(error: error as NSError, adId: adId)
        methodChannel?.invokeMethod("onAdFailedToLoad", arguments: errMap)
    }
    
    func interstitial(_ interstitial: POBInterstitial, didFailToShowAdWithError error: Error) {
        let errMap = POBUtils.convertErrorAndAdIdToMap(error: error as NSError, adId: adId)
        methodChannel?.invokeMethod("onAdFailedToShow", arguments: errMap)
    }
    
    // Notifies the delegate that the interstitial ad will be presented as a modal
    func interstitialWillPresentAd(_ interstitial: POBInterstitial) {
        // no op
    }
    
    func interstitialDidPresentAd(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdOpened", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that the interstitial ad has been animated off the screen.
    func interstitialDidDismissAd(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdClosed", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate of ad click
    func interstitialDidClickAd(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdClicked", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that a user interaction will open another app (e.g. App Store), leaving the current app.
    func interstitialWillLeaveApplication(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAppLeaving", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that the ad has expired
    func interstitialDidExpireAd(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdExpired", arguments: POBUtils.adIdArgs(adId: adId))
    }

    // Notifies the delegate that the video playback is completed
    func interstitialDidFinishVideoPlayback(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onVideoPlaybackCompleted", arguments: POBUtils.adIdArgs(adId: adId))
    }

    // Notifies the delegate that the ad impression
    func interstitialDidRecordImpression(_ interstitial: POBInterstitial) {
        methodChannel?.invokeMethod("onAdImpression", arguments: POBUtils.adIdArgs(adId: adId))
    }
}
